pkg/dogstatsd/listeners/packet_buffer.go

```diff
@@ -1,67 +1,76 @@
-// Unless explicitly stated otherwise all files in this repository are licensed
-// under the Apache License Version 2.0.
-// This product includes software developed at Datadog (https://www.datadoghq.com/).
-// Copyright 2016-2019 Datadog, Inc.
-
 package listeners
 
 import (
 	"sync"
 	"time"
 )
 
-// packetBuffer is a buffer of packet that will automatically flush to configurable channel
-// when it is full or after a configurable duration
+const messageSeparator = byte('\n')
+
+// packetBuffer merges multiple incoming datagrams into one "Packet" object to
+// save space and make number of message in a single "Packet" more predictable
 type packetBuffer struct {
-	packets       Packets
+	packet        *Packet
+	packetLength  int
+	pool          *PacketPool
+	packetsBuffer *packetsBuffer
 	flushTimer    *time.Ticker
-	bufferSize    uint
-	outputChannel chan Packets
 	closeChannel  chan struct{}
-	m             sync.Mutex
+	sync.Mutex
 }
 
-func newPacketBuffer(bufferSize uint, flushTimer time.Duration, outputChannel chan Packets) *packetBuffer {
-	pb := &packetBuffer{
-		bufferSize:    bufferSize,
+func newPacketBuffer(pool *PacketPool, flushTimer time.Duration, packetsBuffer *packetsBuffer) *packetBuffer {
+	packetBuffer := &packetBuffer{
+		packet:        pool.Get(),
+		pool:          pool,
+		packetsBuffer: packetsBuffer,
 		flushTimer:    time.NewTicker(flushTimer),
-		outputChannel: outputChannel,
-		packets:       make(Packets, 0, bufferSize),
 		closeChannel:  make(chan struct{}),
 	}
-	go pb.flushLoop()
-	return pb
+	go packetBuffer.flushLoop()
+	return packetBuffer
 }
 
-func (pb *packetBuffer) flushLoop() {
+func (p *packetBuffer) flushLoop() {
 	for {
 		select {
-		case <-pb.flushTimer.C:
-			pb.m.Lock()
-			pb.flush()
-			pb.m.Unlock()
-		case <-pb.closeChannel:
+		case <-p.flushTimer.C:
+			p.Lock()
+			p.flush()
+			p.Unlock()
+		case <-p.closeChannel:
 			return
 		}
 	}
 }
 
-func (pb *packetBuffer) append(packet *Packet) {
-	pb.m.Lock()
-	defer pb.m.Unlock()
-	pb.packets = append(pb.packets, packet)
-	if uint(len(pb.packets)) == pb.bufferSize {
-		pb.flush()
+func (p *packetBuffer) addMessage(message []byte) {
```

Comment:
That is sad that there is no way (without allocation) to receive a message ended with a `\n` here, it would improve this method reading quality.